### ISSUE TITLE
False positive for class patterns when `Callable` is used

### ISSUE THREAD

**Issue by erictraut** [python/mypy#16617]
Mypy generates a false positive with the following code.

```python
from collections.abc import Callable
from typing import TypeVar

T = TypeVar("T")

def foo(obj: T | Callable[..., T]) -> None:
    match obj:
        case Callable() as func:  # error: Expected type in class pattern; found "typing._SpecialForm"
            func()
```

FWIW, [pyright](https://github.com/microsoft/pyright/issues/6648) was also not handling this case correctly.

**Comment by hauntsaninja** [comment]
Fixed by #19471
